src/libstd/keyword_docs.rs

```diff
@@ -1578,12 +1578,181 @@ mod type_keyword {}
 
 #[doc(keyword = "unsafe")]
 //
-/// Code or interfaces whose [memory safety] cannot be verified by the type system.
+/// Code or interfaces whose [memory safety] cannot be verified by the type
+/// system.
 ///
-/// The documentation for this keyword is [not yet complete]. Pull requests welcome!
+/// The `unsafe` keyword has two uses: to declare the existence of contracts the
+/// compiler can't check (`unsafe fn` and `unsafe trait`), and to declare that a
+/// programmer has checked that these contracts have been upheld (`unsafe {}`
+/// and `unsafe impl`, but also `unsafe fn` -- see below). They are not mutually
+/// exclusive, as can be seen in `unsafe fn`.
+///
+/// Unsafe Rust adds some capabilities to Safe Rust:
+///
+/// - Dereference [raw pointers]
+/// - Call `unsafe` functions (including foreign functions, some compiler
+/// intrinsics and the raw allocator)
+/// - Implement `unsafe` [`trait`]s
+/// - Mutate [`static`]s (including [`extern`]al ones)
+/// - Access fields of [`union`]s
+///
+/// # Why do we need a keyword for this ?
+///
+/// **No matter what, Safe Rust can't cause Undefined Behavior**. This is
+/// referenced to as [soundness]: a well-typed program actually has the desired
+/// properties. In Safe Rust, the language and the compiler do this mostly by
+/// themselves whereas in Unsafe Rust the programmer must help. The [Nomicon]
+/// has a more detailed explanation on the subject.
```

Comment:
Should we also explain the other similar concept: unsafe, unsound, undefined in this page like in a section `soundness`? https://cheats.rs/#unsafe-unsound-undefined

Reply:
I'm not sure about this. I feel like the documentation is already really heavy, I don't know if adding more theory/vocabulary is a good idea for a keyword doc item.